### PR TITLE
Add AVL Tree Implementation in Java and Python

### PR DESCRIPTION
**Pull request by bhushan-borole**
AVL Tree Implementaion in Java and Python.

Operations:
- insert
- delete
- search
- in-order Traversal
- post-order Traversal
- pre-order Traversal
- pretty print

### CODE REVIEW

**Review by vinay-deshmukh — APPROVED** [review]

File: Python/avl.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        self.height = 1
+
+    def __repr__(self):
+        return f'Node(value={self.value}, height={self.height})'
+
+class AVL:
+    '''A class to denote an AVL Tree.'''
+    def __init__(self, args=None):
+        self.root = None
+        if args:
```
Comment (RishabhBhatnagar) [inline Python/avl.py:41]: Any description upon the value of args? Maybe move it into __init__ as an optional parameter.

File: Python/avl.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        node.height = 1 + max(self.get_height(node.left), self.get_height(node.right))
+
+        # Rebalance the node
+        return self._update_balance(node, node.value)
+
+    def in_order(self, node):
+        '''Recursive In Order traversal of the tree'''
+        if node is not None:
+            self.in_order(node.left)
+            print(node.value, end=' ')
```
Comment (RishabhBhatnagar) [inline Python/avl.py]: printing values to stdout is often considered a bad practice. Either return a collection of node values or input a logger from the user where s/he wants the output to be written.

**Review by RishabhBhatnagar — DISMISSED** [review]

File: Python/avl.py (hunk 1 above)
Comment (bhushan-borole) [reply Python/avl.py:41]: Adding the description.

File: Python/avl.py (hunk 2 above)
Comment (bhushan-borole) [reply Python/avl.py]: Okay, will make these changes

File: Java/AVL.java [hunk 3]
```diff
@@ -193,27 +185,28 @@ public AVLNode search(int value) {
...
-        }
+        return node;
+        // if(node != null) {
+        //     System.out.println("Node found");
+        //     return node;
+        // }
+        // else {
+        //     System.out.println("Node not found");
+        //     return null;
+        // }
```
Comment (vinay-deshmukh) [inline Java/AVL.java:196]: Remove unused code

File: Java/AVL.java [hunk 4]
```diff
@@ -193,27 +185,28 @@ public AVLNode search(int value) {
...
-        }
-        else {
-            System.out.println("Node not found");
-            return null;
-        }
+        return node;
+        // if(node != null) {
+        //     System.out.println("Node found");
+        //     return node;
+        // }
```
Comment (vinay-deshmukh) [inline Java/AVL.java]: Remove commented code

File: Java/AVL.java [hunk 5]
```diff
@@ -193,27 +185,28 @@ public AVLNode search(int value) {
...
 
     private AVLNode search(AVLNode node, int value) {
         // Best case: value is present or if node does not exist
         if(node == null || node.value == value)
             return node;
 
-        // Since AVL Tree follows property of BST, if value is lesser than current node's value,
-        // it will be present in left sub tree.
+            // Since AVL Tree follows property of BST, if value is lesser than current node's value,
+            // it will be present in left sub tree.
```
Comment (vinay-deshmukh) [inline Java/AVL.java]: De indent comments

File: Java/AVL.java [hunk 6]
```diff
@@ -193,27 +185,28 @@ public AVLNode search(int value) {
...
         else if(value < node.value)
             return search(node.left, value);
 
-        // If value is greater than current node's value then it will be present in right sub tree.
+            // If value is greater than current node's value then it will be present in right sub tree.
```
Comment (vinay-deshmukh) [inline Java/AVL.java]: De indent comments here as well

**Review by vinay-deshmukh — CHANGES_REQUESTED** [review]

File: Java/AVL.java (hunk 6 above)
Comment (bhushan-borole) [reply Java/AVL.java]: Resolved

File: Java/AVL.java (hunk 5 above)
Comment (bhushan-borole) [reply Java/AVL.java]: resolved

File: Java/AVL.java (hunk 4 above)
Comment (bhushan-borole) [reply Java/AVL.java]: resolved

**Review by RishabhBhatnagar — APPROVED** [review]
Good Work!

**Review by vinay-deshmukh — APPROVED** [review]

**Pull request merged**